File: python/py-wyrd/python/examples/workflow_structured_output.py
"""Two-step structured-output workflow against the live OpenAI API."""

from __future__ import annotations

from pydantic import BaseModel

from wyrd import Agent, Prompt, Workflow


class Plan(BaseModel):
    summary: str
    steps: list[str]


def main() -> None:
    planner = Agent(
        prompt=Prompt(
            provider="openai",
            model="gpt-5.4-nano-2026-03-17",
            system="You produce structured plans.",
            messages=["Plan: ${topic}"],
            output=Plan,
        ),
        name="planner",
    )
    writer = Agent(
        prompt=Prompt(
            provider="openai",
            model="gpt-5.4-nano-2026-03-17",
            system="You write 2-sentence briefs.",
            messages=["Write a brief from this summary: ${summary}"],
        ),
        name="writer",
    )
    wf = Workflow.sequential("demo", planner, writer)
    run = wf.run({"topic": "the Rust borrow checker"})
    print("parameters:", run.parameters)
    print("final:", run.final_output)


if __name__ == "__main__":
    main()

File: crates/wyrd/wyrd/examples/workflow_structured_output.rs
//! Two-step structured-output workflow against the live OpenAI API.
//!
//! Run with:
//! `cargo run -p wyrd --example workflow_structured_output --all-features`.

use serde_json::json;
use wyrd::agent::{Agent, OpenAiChatOptions, Workflow, openai_chat};
use wyrd::prompt::ResponseFormat;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    wyrd::init();

    let planner_schema = json!({
        "type": "object",
        "properties": {
            "summary": {"type": "string"},
            "steps": {"type": "array", "items": {"type": "string"}}
        },
        "required": ["summary", "steps"],
        "additionalProperties": false
    });

    let planner = Agent::new(openai_chat(
        "gpt-5.4-nano-2026-03-17",
        OpenAiChatOptions {
            system: Some("You produce structured plans.".into()),
            messages: vec!["Plan: ${topic}".into()],
            output: Some(ResponseFormat::json_schema("plan", planner_schema)?),
            ..OpenAiChatOptions::default()
        },
    )?)
    .name("planner");

    let writer = Agent::new(openai_chat(
        "gpt-5.4-nano-2026-03-17",
        OpenAiChatOptions {
            system: Some("You write 2-sentence briefs.".into()),
            messages: vec!["Write a brief from this summary: ${summary}".into()],
            ..OpenAiChatOptions::default()
        },
    )?)
    .name("writer");

    let wf = Workflow::sequential("demo", [planner, writer])?;
    let run = wf.run(json!({"topic": "the Rust borrow checker"})).await?;

    println!("parameters: {:#?}", run.parameters);
    println!("final: {}", run.final_output.as_deref().unwrap_or("(none)"));
    Ok(())
}
